### PR TITLE
[api2] stat_file

### PR DESCRIPTION
**Pull request by zMingGit**

### CODE REVIEW

File: seahub/api2/endpoints/admin/sysinfo.py [hunk 1]
```diff
@@ -153,3 +163,121 @@ def get(self, request, format=None):
...
+        end_time = request.GET.get("end", "")
+        group_by = request.GET.get("group_by", "hour")
+        if not start_time:
+            error_msg = _("Start time can not be empty")
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+        if not end_time:
+            error_msg = _("End time can not be empty")
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+        if group_by.lower() not in ["hour", "day"]:
+            error_msg = "Record only can group by day or hour"
```
Comment (freeplant) [inline seahub/api2/endpoints/admin/sysinfo.py]: Records can only group by day or hour

File: seahub/api2/endpoints/admin/sysinfo.py [hunk 2]
```diff
@@ -82,11 +91,11 @@ def get(self, request, format=None):
...
             inactive_ldap_users = 0
 
-        active_users = active_db_users + active_ldap_users if active_ldap_users > 0 \
-                else active_db_users
+        active_users = active_db_users + active_ldap_users if \
+            active_ldap_users > 0 else active_db_users
 
-        inactive_users = inactive_db_users + inactive_ldap_users if inactive_ldap_users > 0 \
-                else inactive_db_users
+        inactive_users = inactive_db_users + inactive_ldap_users if\
```
Comment (freeplant) [inline seahub/api2/endpoints/admin/sysinfo.py]: if \, need space

File: seahub/api2/endpoints/admin/sysinfo.py [hunk 3]
```diff
@@ -153,3 +162,137 @@ def get(self, request, format=None):
...
+        end_time = request.GET.get("end", "")
+        group_by = request.GET.get("group_by", "hour")
+        if not start_time:
+            error_msg = "Start time can not be empty"
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+        if not end_time:
+            error_msg = "End time can not be empty"
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+        if group_by.lower() not in ["hour", "day"]:
+            error_msg = "Record can only group by day or hour"
```
Comment (llj) [inline seahub/api2/endpoints/admin/sysinfo.py]: mark

File: seahub/api2/endpoints/admin/sysinfo.py [hunk 4]
```diff
@@ -153,3 +162,137 @@ def get(self, request, format=None):
...
+                                                  "%Y-%m-%d %H:%M:%S")
+        except:
+            error_msg = "End time %s invalid" % end_time
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+
+        return func(view, request, start_time, end_time, group_by)
+    return _decorated
+
+
+def get_data_by_hour_or_day_by_hour_or_day(parameter, func, func_by_day):
```
Comment (llj) [inline seahub/api2/endpoints/admin/sysinfo.py]: check the function name.

File: seahub/api2/endpoints/admin/sysinfo.py [hunk 5]
```diff
@@ -153,3 +162,137 @@ def get(self, request, format=None):
...
+
+        res_data = []
+        dict_data = {}
+        for i in data:
+            timestamp = str(int(time.mktime(i[0].timetuple())))
+            if dict_data.get(timestamp, None) is None:
+                dict_data[timestamp] = {}
+            dict_data[timestamp][i[1]] = i[2]
+        if len(dict_data) == 0:
+            return Response(status.HTTP_200_OK)
```
Comment (llj) [inline seahub/api2/endpoints/admin/sysinfo.py]: mark

File: seahub/urls.py [hunk 6]
```diff
@@ -50,7 +50,7 @@
 from seahub.api2.endpoints.admin.file_audit import FileAudit
 from seahub.api2.endpoints.admin.file_update import FileUpdate
 from seahub.api2.endpoints.admin.perm_audit import PermAudit
-from seahub.api2.endpoints.admin.sysinfo import SysInfo
+from seahub.api2.endpoints.admin.sysinfo import SysInfo, FileOperationsView, TotalStorageView, ActiveUsersView
```
Comment (llj) [inline seahub/urls.py]: mark

File: seahub/urls.py [hunk 7]
```diff
@@ -240,6 +240,9 @@
 
     ## admin::sysinfo
     url(r'^api/v2.1/admin/sysinfo/$', SysInfo.as_view(), name='api-v2.1-sysinfo'),
+    url(r'^api/v2.1/admin/sysinfo/file-operations/$', FileOperationsView.as_view(), name='api-v2.1-admin-file-operations'),
+    url(r'^api/v2.1/admin/sysinfo/total-storage/$', TotalStorageView.as_view(), name='api-v2.1-admin-total-storage'),
+    url(r'^api/v2.1/admin/sysinfo/active-users/$', ActiveUsersView.as_view(), name='api-v2.1-admin-active-users'),
```
Comment (llj) [inline seahub/urls.py]: mark.

File: seahub/api2/endpoints/admin/statistics.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        Get records of the specified time range.
+            param:
+                start: the start time of the query.
+                end: the end time of the query.
+                group_by: group records by day or by hour, default group by hour.
+            return:
+                the list of file operations record.
+        """
+        data = get_data_by_hour_or_day(group_by, start_time, end_time, get_file_ops_stats, get_file_ops_stats_by_day)
+        if data is None:
```
Comment (xiez) [inline seahub/api2/endpoints/admin/statistics.py]: 返回不会为none

File: seahub/api2/endpoints/admin/statistics.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        end_time = request.GET.get("end", "")
+        group_by = request.GET.get("group_by", "hour")
+        if not start_time:
+            error_msg = "Start time can not be empty"
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+        if not end_time:
+            error_msg = "End time can not be empty"
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+        if group_by.lower() not in ["hour", "day"]:
+            error_msg = "Records can only group by day or hour"
```
Comment (xiez) [inline seahub/api2/endpoints/admin/statistics.py]: group_by can only be day or hour

File: seahub/api2/endpoints/admin/statistics.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+            error_msg = "End time %s invalid" % end_time
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+
+        return func(view, request, start_time, end_time, group_by)
+    return _decorated
+
+
+def get_data_by_hour_or_day(parameter, start_time, end_time, func, func_by_day):
+    timezone_name = timezone.get_current_timezone_name()
+    offset = pytz.timezone(timezone_name).localize(datetime.datetime(2011,1,1)).strftime('%z')
```
Comment (xiez) [inline seahub/api2/endpoints/admin/statistics.py]: 2011 改称 now

File: seahub/api2/endpoints/admin/statistics.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    offset = offset[:3] + ':' + offset[3:]
+    if parameter == "hour":
+        data = func(start_time, end_time, offset)
+        logger.error("seafile hour:%s" % str(data))
```
Comment (xiez) [inline seahub/api2/endpoints/admin/statistics.py]: 调试信息可以去掉了

Comment (zMingGit) [inline seahub/api2/endpoints/admin/statistics.py]: 好

**Pull request merged**